[mlir][arith] Fix overflow bug in arith::CeilDivSIOp::fold

banach-space opened the pull request:
The folder for arith::CeilDivSIOp should only be applied when it can be
guaranteed that no overflow would happen. The current implementation
works fine when both dividends are positive and the only arithmetic
operation is the division itself.

However, in cases where either the dividend or divisor is negative  (or both),
the division is split into multiple arith operations, e.g.: `- ( -a / b)`. That's
additional 2 operations on top of the actual division that can overflow 
- the folder should check all 3 ops for overflow. 

The current logic doesn't do that - it effectively only checks the last operation
(i.e. the division). It breaks when using e.g. MININT values (e.g. -128 for
8-bit integers) - negating such values overflows.

This PR makes sure that no folding happens if any of the intermediate
arithmetic operations overflows.

Fixes https://github.com/llvm/llvm-project/issues/89382

banach-space committed:
[mlir][arith] Fix overflow bug in arith::CeilDivSIOp::fold

The folder for arith::CeilDivSIOp should only be applied when it can be
guaranteed that no overflow would happen. The current implementation
works fine when both dividends are positive and the only arithmetic
operation is the division itself.

However, in cases where at least one of the dividends is negative, the
division is split into multiple operations, e.g.: `- ( -a / b)`. That's
additional 2 operations on top of the actual division that can overflow
- the folder should check all 3 ops for overflow. The current logic
doesn't do that - it effectively only the last operation (i.e. the
division). It breaks when using e.g. MININT values (e.g. -128 for
8-bit integers) - negating such values overflows.

This PR makes sure that no folding happens if any of the intermediate
arithmetic operations overflows.

Lewuathe commented on mlir/lib/Dialect/Arith/IR/ArithOps.cpp:
```diff
@@ -701,22 +701,34 @@ OpFoldResult arith::CeilDivSIOp::fold(FoldAdaptor adaptor) {
         }
+
+        bool overflowNegA = false;
+        bool overflowNegB = false;
```
The name should have `OrDiv0` at the end more precisely?

```suggestion
        bool overflowNegBOrDiv0 = false;
```

banach-space replied:
Not quite, it's more like this:
```cpp
         bool overflowNegA = false;       // Overflow for "-a"
         bool overflowNegB = false;       // Overflow for "-b"
         bool overflowDiv = false;        // Overflow for "div(lhs, rhs)"
         bool overflowNegDiv = false;     // Overflow for "-div(..., ...)"
```

As in, there's a separate flag for every arithmetic operation that may follow. Also, at this point we already know that `b != 0`.

I might be missing sth, so please correct me if I'm wrong.

banach-space commented on mlir/lib/Dialect/Arith/IR/ArithOps.cpp:
```diff
@@ -701,22 +701,34 @@ OpFoldResult arith::CeilDivSIOp::fold(FoldAdaptor adaptor) {
           return signedCeilNonnegInputs(a, b, overflowOrDiv0);
         }
+
+        bool overflowNegA = false;
```
Good suggestion, thanks!

banach-space committed:
fixup! [mlir][arith] Fix overflow bug in arith::CeilDivSIOp::fold

Add comment, update var name (overflowNegDiv -> overflowNegRes)

banach-space commented:
Thanks for taking a look @Lewuathe 🙏🏻

Lewuathe commented on mlir/test/Transforms/constant-fold.mlir:
```diff
@@ -478,6 +478,32 @@ func.func @simple_arith.ceildivsi() -> (i32, i32, i32, i32, i32) {
+  // overflows). Hence folding should not happen and the original ceildivsi is
+  // preserved.
+
+  // CHECK-COUNT-3:  arith.ceildivsi
```
It's nit. It might have been better to check whether the original one is preserved instead of just checking the count.

something like

```suggestion
  // CHECK-NEXT: %[[C1:.*]] = arith.constant 7 : i8
  // CHECK-NEXT: %[[MIN1:.*]] = arith.constant -128 : i8
  // CHECK-NEXT: %[[CEILDIV1:.*]] = arith.ceildivsi %[[MIN1]], %[[C1]]  : i32
```

banach-space replied:
Happy to update this, no harm in being more thorough.

Lewuathe reviewed (approved):
I think the approach talked in the issue seems correct. Fixing the overflow issue makes sense. Although I have left the one minor comment about the test, it LGTM. 

https://github.com/llvm/llvm-project/issues/89382#issuecomment-2093779774

banach-space committed:
fixup! fixup! [mlir][arith] Fix overflow bug in arith::CeilDivSIOp::fold

Harden the test

kuhar reviewed (commented):
Do we need to update the wording around overflow conditions in the op docs? https://mlir.llvm.org/docs/Dialects/ArithOps/#arithceildivsi-arithceildivsiop this doesn't define what we understand by signed division overflow

pingshiyu commented:
I'm a little confused here (mostly due to my own ignorance): the original bug was on the conversion pass `--convert-arith-to-llvm`, but this patch is on the constant folder. Just wondering if you could point out where in the codebase is the constant folder called duirng conversion?

Perhaps an extra integration test, consisting of the original file with `arith-expand` or `convert-arith-to-llvm` pass, would give a bit further assurance?

joker-eph commented:
>  Just wondering if you could point out where in the codebase is the constant folder called duirng conversion?

Constant folding is always involved by the dialect conversion framework implicitly.

> Perhaps an extra integration test, consisting of an end-to-end run of the original file with arith-expand or convert-arith-to-llvm pass, would give a bit further assurance or this patch?

This shouldn't be needed: when we root cause the problem we focus on unit-testing instead.

joker-eph commented on mlir/test/Transforms/constant-fold.mlir:
```diff
@@ -478,6 +478,41 @@ func.func @simple_arith.ceildivsi() -> (i32, i32, i32, i32, i32) {
+// CHECK-LABEL: func @simple_arith.ceildivsi_overflow
+func.func @simple_arith.ceildivsi_overflow() -> (i8, i16, i32) {
+  // The negative values below are MININTs for the corresponding bit-width. The
+  // folder will try to negate them (so that the division operartes on two
```
```suggestion
  // folder will try to negate them (so that the division operates on two
```

banach-space replied:
Thanks!

banach-space committed:
fixup! fixup! fixup! [mlir][arith] Fix overflow bug in arith::CeilDivSIOp::fold

Fix typo

joker-eph reviewed (approved):
This seems to me like it makes the folder strictly more conservative, which should be safe.

It is still open whether we couldn't be more aggressive though: are there cases where we could fold but we don't right now? (if so can you add a TODO? Possibly a test for these?) 
Can we accurately catch the real overflow cases and folds to poison for example? (arith may not generate poison out-of-thin air for at the moment, which could be a blocker)

banach-space committed:
fixup! fixup! fixup! fixup! [mlir][arith] Fix overflow bug in arith::CeilDivSIOp::fold

Add comments

banach-space commented:
> Do we need to update the wording around overflow conditions in the op docs? https://mlir.llvm.org/docs/Dialects/ArithOps/#arithceildivsi-arithceildivsiop this doesn't define what we understand by signed division overflow

+1, but I'd rather defer to somebody with a stronger opinion and more experience with Arith.

> It is still open whether we couldn't be more aggressive though

Yes, we could be more aggressive. In fact, that's what @bviyer is proposing here:
* https://github.com/llvm/llvm-project/pull/90855

I suggested that first we make sure that we don't drop these overflow flags on the floor (i.e. that we land this PR first).

>  are there cases where we could fold but we don't right now? (if so can you add a TODO? Possibly a test for these?)

Yes, the cases in the test that I added (I've added TODOs). There might be more, but nothing obvious comes to mind. I don't see that many edge cases being tested for Arith - there might be even more room for improvement.

banach-space commented:
I'll land this later today. 

> Do we need to update the wording around overflow conditions in the op docs?

@kuhar, please let me know if you feel that this deserves a note in the docs. My intention was to refine/fix the implementation details rather than change the Op semantics. But I appreciate that in practice one defines the other (not sure whether that's intended though).

kuhar commented:
> @kuhar, please let me know if you feel that this deserves a note in the docs. My intention was to refine/fix the implementation details rather than change the Op semantics. But I appreciate that in practice one defines the other (not sure whether that's intended though).

Yeah I think adding a few examples of overflowing inputs there would go a long way and won't take much time to write.

banach-space committed:
fixup! fixup! fixup! fixup! fixup! [mlir][arith] Fix overflow bug in arith::CeilDivSIOp::fold

Add comments

banach-space commented:
> Yeah I think adding a few examples of overflowing inputs there would go a long way and won't take much time to write.

Turns out this is already documented, so I only added a note documenting the changes made in this patch. @kuhar , wdyt?

joker-eph commented on mlir/include/mlir/Dialect/Arith/IR/ArithOps.td:
```diff
@@ -587,10 +587,11 @@ def Arith_CeilDivSIOp : Arith_IntBinaryOp<"ceildivsi",
-    signed division overflow.
+    Divison by zero, or signed division overflow (minimum value divided by -1)
+    is undefined behavior. While dividing minimum value by a value != -1 shouldn't
+    overflow, the current implementation treats it as such. When applied to `vector`
```
The sentence makes it look like dividing minimum value by a value != -1 is undefined behavior, I don't believe this is the intent.

I am not sure what you're trying to document here? Implementation details of the folder aren't relevant to the op semantics I believe.

kuhar commented:
@banach-space I wanted to clarify whether what we have there currently is the only kind of signed division overflow or if there are more (which I thought may be the case looking at the fold implementation);
> Divison by zero, or signed division overflow (minimum value divided by -1)

joker-eph commented:
> the fold should handle all defined cases in the dialect doc

This is the TODO we discussed before I believe.

> we either need to make the semantics less defined 

Why? It seems just fine to have "incomplete" or "conservative" folders: these are opportunistic, not guaranteed to fold all possible cases in general.

kuhar commented:
> Why? It seems just fine to have "incomplete" or "conservative" folders: these are opportunistic, not guaranteed to fold all possible cases in general.

Yes, I think we agree on this. That's what I meant by 'should' (vs. 'must').

banach-space commented:
> Implementation details of the folder aren't relevant to the op semantics I believe.

Agreed. It wasn't clear to me what the expectation was, so I proposed sth to progress the discussion. Like you said, it's an implementation detail, so let me revert that.

>  my only concern is we are discovering more overflow cases that don't lower / fold properly yet are allowed by the op spec.

The cases that's being fixed here is not  `arith.ceildivsi` overflow. Also, based on the the spec, I assume that we don't really care how overflow cases are lowered/folded (additional emphasis from me):

> Divison by zero, or signed division overflow (minimum value divided by -1) is **undefined behavior**. 

I feel that we are all on the same page here, so I will land this later today. Please let me know if you feel differently.

banach-space committed:
fixup! [mlir][arith] Fix overflow bug in arith::CeilDivSIOp::fold"

This reverts commit a9e8f3a436719c44ab5cc3b3ac6a7198d416cb4c (the
previous commit)
